File: LeetCode/Interview/662二叉树的最大宽度/width.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def widthOfBinaryTree(self, root: TreeNode) -> int:
        if not root:
            return 0

        # 初始化队列，存储节点及其对应的索引
        queue = deque([(root, 0)])
        max_width = 0

        # 开始 BFS 遍历
        while queue:
            level_length = len(queue)
            level_min = queue[0][1]  # 该层最左节点的索引
            for i in range(level_length):
                node, index = queue.popleft()

                # 记录当前层的宽度
                if node.left:
                    queue.append((node.left, 2 * index))
                if node.right:
                    queue.append((node.right, 2 * index + 1))

            # 计算当前层的宽度
            level_max = queue[-1][1] if queue else index
            max_width = max(max_width, level_max - level_min + 1)

        return max_width

File: LeetCode/Interview/662二叉树的最大宽度/tree_width/src/width.c
/******************************************************************************
 *
 * Name: width.c - Description
 * Created on 2025/02/25
 * Copyright (C) 2022 - 2025, wyh.
 *
 *****************************************************************************/

#include "width.h"
#include "queue.h"
#include <stdio.h>
#include <stdlib.h>

static int queue_size(Queue *q)
{
	if (q == NULL || q->front == NULL)
		return 0;

	int count = 0;
	QueueNode *tmp = q->front;
	while (tmp) {
		count++;
		tmp = tmp->next;
	}

	return count;
}

int widthOfBinaryTree(TreeNode *root)
{
	if (!root)
		return 0;

	Queue *queue = create_queue();
	enqueue(queue, root, 0);
	int max_width = 0;

	while (!is_empty(queue)) {
		int size = queue_size(queue);
		unsigned long min_index = front(queue)->index;
		unsigned long first, last;

		for (int i = 0; i < size; i++) {
			QueueNode *q_node = front(queue);
			dequeue(queue);
			TreeNode *t_node = q_node->node;
			unsigned long index = q_node->index - min_index;

			if (i == 0)
				first = index;
			if (i == size - 1)
				last = index;

			if (t_node->left)
				enqueue(queue, t_node->left, 2 * index);
			if (t_node->right)
				enqueue(queue, t_node->right, 2 * index + 1);
		}

		max_width = (last - first + 1) > max_width ?
				    (last - first + 1) :
				    max_width;
	}

	free_queue(queue);
	return max_width;
}
